### PR TITLE
Adding Homepage and other functionalities

### PR DESCRIPTION
**Pull request by AbdulBasit0044**
In this commit I have added the following components, pages and hooks:

- Homepage
- Create Movie
- Delete Movie
- Movie Details
- Movie (which generate cards by fetching data from the api)
- userFetch (this is the custom hook to fetch data from the api)

### CODE REVIEW

File: src/App.jsx [hunk 1]
```diff
@@ -1,22 +1,24 @@
...
 function App() {
   return (
     <Router>
       <div className="App">
         <Navbar />
         <Routes>
-          <Route path="" element={<Homepage />} />
-          <Route path="about" element={<About />} />
-          <Route path="create" element={<Create />} />
+          <Route exact path="" element={<Homepage />} />
```
Comment (omerzn) [inline src/App.jsx]: why did you add `exact` again?

File: src/components/homepage/Homepage.jsx [hunk 2]
```diff
@@ -1,8 +1,17 @@
+import { useState, useEffect } from 'react'
+import Movie from '../movie/Movie'
+import useFetch from '../../hooks/useFetch';
+import { MovieContainer } from './Styles/StyleMovieContainer';
+
 const Homepage = () => {
+    const url = 'http://localhost:8001/movies';
```
Comment (omerzn) [inline src/components/homepage/Homepage.jsx]: make env file and add this `http://localhost:8001` there. After that create an API folder with movies.js file: `src/API/movies.js` Create a function called getMovies() and fetch the data. Use the env variable there and the rest of the url. Import that function here and use it.

File: src/components/homepage/Homepage.jsx [hunk 3]
```diff
@@ -1,8 +1,17 @@
...
+    const { movies, isPending, error } = useFetch(url);
     return (
-        <div className="homepage">
-            <p>This is home page</p>
-        </div>
+        <MovieContainer>
+            <div className="movie-container">
```
Comment (omerzn) [inline src/components/homepage/Homepage.jsx]: this div is redundant. `MovieContainer` is already a div. Please remove this and move the required styling to the `MovieContainer` div.

File: src/components/movie/Movie.jsx [hunk 4]
```diff
@@ -0,0 +1,23 @@
+import { Link } from "react-router-dom";
+
+const Movie = ({ movies }) => {
+  return (
+    <>
```
Comment (omerzn) [inline src/components/movie/Movie.jsx:5]: is this necessary? Does it give an error without this?

File: src/components/movie/Movie.jsx [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+      {movies.map((movie) => (
+        <Link to={`/movies/${movie.id}`}>
```
Comment (omerzn) [inline src/components/movie/Movie.jsx]: the key prop should given here

File: src/components/movie/Movie.jsx [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+          <div className="movie">
+            <div className="movie-img" key={movie.id}>
```
Comment (omerzn) [inline src/components/movie/Movie.jsx]: not here

File: src/components/movie/MovieDetails.jsx [hunk 7]
```diff
@@ -0,0 +1,33 @@
+import useFetch from "../../hooks/useFetch";
+import { useNavigate, useParams } from "react-router-dom";
+
+const MovieDetails = () => {
+    const { id } = useParams();
+    const navigate = useNavigate();
+    console.log("params are ", id);
```
Comment (omerzn) [inline src/components/movie/MovieDetails.jsx]: remove console

File: src/components/movie/MovieDetails.jsx [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+    const url = 'http://localhost:8001/movies/';
```
Comment (omerzn) [inline src/components/movie/MovieDetails.jsx]: same as mentioned above

File: src/components/movie/MovieDetails.jsx [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+    const { movies: movie, isPending, error } = useFetch(url + id)
+
+    const handleDelete = () => {
+        fetch('http://localhost:8001/movies/' + movie.id, {
```
Comment (omerzn) [inline src/components/movie/MovieDetails.jsx]: same

File: src/components/navbar/Create.jsx [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+  const [rating, setRating] = useState("");
+  const [img, setImg] = useState(
+    "https://i.ytimg.com/vi/VzvdoLeXClg/movieposter_en.jpg"
+  );
+  const navigate = useNavigate();
+
+  const handleAddMovieSubmit = (e) => {
+    e.preventDefault();
+    const movie = { title, director, overview, rating, img };
+    console.log(movie);
```
Comment (omerzn) [inline src/components/navbar/Create.jsx]: remove

File: src/components/navbar/Create.jsx [hunk 11]
```diff
@@ -0,0 +1,71 @@
...
+    fetch("http://localhost:8001/movies/", {
```
Comment (omerzn) [inline src/components/navbar/Create.jsx]: same as mentioned above

File: src/components/navbar/Create.jsx [hunk 12]
```diff
@@ -0,0 +1,71 @@
...
+      method: "POST",
+      headers: { "Content-Type": "application/json" },
+      body: JSON.stringify(movie),
+    }).then(() => {
```
Comment (omerzn) [inline src/components/navbar/Create.jsx]: dont move this to the api file, redirection should be handled here

File: src/components/navbar/Create.jsx [hunk 13]
```diff
@@ -0,0 +1,71 @@
...
+      navigate("/");
+    });
+  };
+  return (
+    <CreateStyleWrapper>
+      <div className="create">
```
Comment (omerzn) [inline src/components/navbar/Create.jsx]: redundant div

File: src/hooks/useFetch.jsx [hunk 14]
```diff
@@ -0,0 +1,26 @@
...
+    useEffect(() => {
+        fetch(url)
+            .then(resp => {
+                if (!resp.ok) {
+                    throw Error('could not fetch the data for that resource');
+                }
+                return resp.json();
+            })
+            .then(data => {
+                console.log(data);
```
Comment (omerzn) [inline src/hooks/useFetch.jsx]: remove this

**Review by omerzn — CHANGES_REQUESTED** [review]
please go through the requested changes

File: src/App.jsx (hunk 1 above)
Comment (AbdulBasit0044) [reply src/App.jsx]: I reverted the commit from the last branch that is why it got again in this code. It is updated now.

File: src/components/homepage/Homepage.jsx (hunk 3 above)
Comment (AbdulBasit0044) [reply src/components/homepage/Homepage.jsx]: Done

File: src/components/movie/Movie.jsx (hunk 4 above)
Comment (AbdulBasit0044) [reply src/components/movie/Movie.jsx:5]: Yes, it gives an error with this because in return statement we can return only single element.

File: src/components/movie/Movie.jsx (hunk 5 above)
Comment (AbdulBasit0044) [reply src/components/movie/Movie.jsx]: Done

File: src/components/navbar/Create.jsx (hunk 10 above)
Comment (AbdulBasit0044) [reply src/components/navbar/Create.jsx]: Removed

File: src/components/navbar/Create.jsx (hunk 11 above)
Comment (AbdulBasit0044) [reply src/components/navbar/Create.jsx]: Removed

File: src/components/navbar/Create.jsx (hunk 13 above)
Comment (AbdulBasit0044) [reply src/components/navbar/Create.jsx]: Removed

File: src/hooks/useFetch.jsx (hunk 14 above)
Comment (AbdulBasit0044) [reply src/hooks/useFetch.jsx]: Removed

File: src/components/navbar/Create.jsx (hunk 12 above)
Comment (AbdulBasit0044) [reply src/components/navbar/Create.jsx]: Done

File: src/components/homepage/Homepage.jsx (hunk 2 above)
Comment (AbdulBasit0044) [reply src/components/homepage/Homepage.jsx]: Done

File: src/components/homepage/Homepage.jsx [hunk 15]
```diff
@@ -1,9 +1,10 @@
...
+
 const Homepage = () => {
-    return (
-        <div className="homepage">
-            <p>This is home page</p>
-        </div>
-    );
-}
+  const { movies, isPending, error } = GetMovies();
+  return <MovieContainer>{movies && <Movie movies={movies} />}</MovieContainer>;
```
Comment (omerzn) [inline src/components/homepage/Homepage.jsx]:
```suggestion
  return (
  <MovieContainer>
     {movies && <Movie movies={movies} />}
  </MovieContainer>);
```

Comment (omerzn) [reply src/components/homepage/Homepage.jsx]: Proper formatting is important because it makes the code more readable.

File: src/components/movie/MovieDetails.jsx [hunk 16]
```diff
@@ -0,0 +1,28 @@
+import { useNavigate, useParams } from "react-router-dom";
+import { GetMoviesBasedOnId, DeleteMovie } from "../../API/movies";
+
+const MovieDetails = () => {
+  const { id } = useParams();
+  const navigate = useNavigate();
+  const { movies: movie, isPending, error } = GetMoviesBasedOnId(id);
+  const handleDelete = async () => {
+    await DeleteMovie(movie.id).then((resp) => {
+      if (resp.ok) navigate("/");
```
Comment (omerzn) [inline src/components/movie/MovieDetails.jsx]:
```suggestion
      resp?.ok && navigate("/");
```
if res.ok is undefined, it won't give an error because used ? for optional chaining. It is also known as the elvis operator.

File: src/components/movie/MovieDetails.jsx [hunk 17]
```diff
@@ -0,0 +1,28 @@
...
+    });
+  };
+  return (
+    <div className="movie-details">
+      {movie && (
+        <article>
+          <h2>Title {movie.title}</h2>
```
Comment (omerzn) [inline src/components/movie/MovieDetails.jsx:17]: Try to add optional chaining here as well. It's helps us avoid code breakages. Use it especially for values that can be null or undefined. What if there is not a movie.img in the db. Your frontend should have a fallback in place starting with this.

File: src/components/navbar/Create.jsx [hunk 18]
```diff
@@ -0,0 +1,64 @@
...
+const Create = () => {
+  const [title, setTitle] = useState("");
+  const [director, setDirector] = useState("");
+  const [overview, setOverview] = useState("");
+  const [rating, setRating] = useState("");
+  const [img, setImg] = useState(
+    "https://i.ytimg.com/vi/VzvdoLeXClg/movieposter_en.jpg"
+  );
+  const navigate = useNavigate();
+  const handleAddMovieSubmit = async (e) => {
```
Comment (omerzn) [inline src/components/navbar/Create.jsx:16]: add one line space before this function

File: src/components/navbar/Create.jsx [hunk 19]
```diff
@@ -0,0 +1,64 @@
...
+    e.preventDefault();
+    const movie = { title, director, overview, rating, img };
+    await CreateMovie(movie).then((resp) => {
+      if (resp.ok) navigate("/");
```
Comment (omerzn) [inline src/components/navbar/Create.jsx]: same change here

**Review by omerzn — CHANGES_REQUESTED** [review]
minor changes requested

File: src/components/homepage/Homepage.jsx (hunk 15 above)
Comment (AbdulBasit0044) [reply src/components/homepage/Homepage.jsx]: Done

File: src/components/movie/MovieDetails.jsx (hunk 16 above)
Comment (AbdulBasit0044) [reply src/components/movie/MovieDetails.jsx]: This is good thing to learn.Code Updated.

File: src/components/movie/MovieDetails.jsx (hunk 17 above)
Comment (AbdulBasit0044) [reply src/components/movie/MovieDetails.jsx:17]: Done, I have made these fields required to add in the front end.

File: src/components/navbar/Create.jsx (hunk 18 above)
Comment (AbdulBasit0044) [reply src/components/navbar/Create.jsx:16]: Done

File: src/components/navbar/Create.jsx (hunk 19 above)
Comment (AbdulBasit0044) [reply src/components/navbar/Create.jsx]: Updated

**Review by omerzn — APPROVED** [review]
